src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,1054 @@
+                                                                    const std::vector<bool>& right) const {
+  const auto result_size = _result_size(left.size(), right.size());
+
+  if (result_size == 0) return {};
+
+  if (left.size() == right.size()) {
+    std::vector<bool> nulls(result_size);
+    std::transform(left.begin(), left.end(), right.begin(), nulls.begin(), [](auto l, auto r) { return l || r; });
+    return nulls;
+  } else if (left.size() > right.size()) {
+    DebugAssert(right.size() == 1,
+                "Operand should have either the same row count as the other or 1 (to represent a literal)");
+    if (right.front())
+      return std::vector<bool>({true});
+    else
+      return left;
+  } else {
+    DebugAssert(left.size() == 1,
+                "Operand should have either the same row count as the other or 1 (to represent a literal)");
+    if (left.front())
+      return std::vector<bool>({true});
+    else
+      return right;
+  }
+}
+
+void ExpressionEvaluator::_ensure_column_materialization(const ColumnID column_id) {
+  Assert(_chunk, "Cannot access Columns in this Expression as it doesn't operate on a Table/Chunk");
+
+  if (_column_materializations[column_id]) return;
+
+  const auto& column = *_chunk->get_column(column_id);
+
+  resolve_data_type(column.data_type(), [&](const auto column_data_type_t) {
+    using ColumnDataType = typename decltype(column_data_type_t)::type;
+
+    std::vector<ColumnDataType> values;
+    materialize_values(column, values);
+
+    if (_table->column_is_nullable(column_id)) {
+      std::vector<bool> nulls;
+      materialize_nulls<ColumnDataType>(column, nulls);
+      _column_materializations[column_id] =
+          std::make_shared<ExpressionResult<ColumnDataType>>(std::move(values), std::move(nulls));
+
+    } else {
+      _column_materializations[column_id] = std::make_shared<ExpressionResult<ColumnDataType>>(std::move(values));
+    }
+  });
+}
+
+std::shared_ptr<ExpressionResult<std::string>> ExpressionEvaluator::_evaluate_substring(
+    const std::vector<std::shared_ptr<AbstractExpression>>& arguments) {
+  Assert(arguments.size() == 3, "SUBSTR expects three arguments");
+
+  const auto strings = evaluate_expression_to_result<std::string>(*arguments[0]);
+  const auto starts = evaluate_expression_to_result<int32_t>(*arguments[1]);
+  const auto lengths = evaluate_expression_to_result<int32_t>(*arguments[2]);
+
+  const auto row_count = _result_size(strings->size(), starts->size(), lengths->size());
+
+  std::vector<std::string> result_values(row_count);
+  std::vector<bool> result_nulls(row_count);
+
+  for (auto chunk_offset = ChunkOffset{0}; chunk_offset < row_count; ++chunk_offset) {
+    result_nulls[chunk_offset] =
+        strings->null(chunk_offset) || starts->null(chunk_offset) || lengths->null(chunk_offset);
+
+    const auto& string = strings->value(chunk_offset);
+    Assert(string.size() < std::numeric_limits<int32_t>::max(),
```

Comment:
This really could be a `DebugAssert`. Let's try and keep strong asserts out of inner loops.

Reply:
absolutely with you, this just slipped in.